torch/nn/utils/prune.py

```diff
@@ -0,0 +1,1269 @@
+                )
+                + " Found '{}'".format(method._tensor_name)
+            )
+        # if all checks passed, add to _pruning_methods tuple
+        self._pruning_methods += (method,)
+
+    def __len__(self):
+        return len(self._pruning_methods)
+
+    def __iter__(self):
+        return iter(self._pruning_methods)
+
+    def __getitem__(self, idx):
+        return self._pruning_methods[idx]
+
+    def compute_mask(self, t, default_mask):
+        def _combine_masks(method, t, mask):
+            r"""Compute new cumulative mask from old mask * new partial mask.
+            The new partial mask should be computed on the entries that
+            were not zeroed out by the old mask. 
+            Which portions of the tensor the new mask will be calculated from 
+            depends on the PRUNING_TYPE (handled by the type handler):
+            for 'unstructured', the mask will be computed from the raveled list
+            of nonmasked entries;
+            for 'structured', the mask will be computed from the nonmasked
+            channels in the tensor;
+            for 'global', the mask will be computed across all entries.
+
+            Args:
+                method (a BasePruningMethod subclass): pruning method
+                    currently being applied.
+                t (torch.Tensor): tensor representing the parameter to prune
+                    (of same dimensions as mask).
+                mask (torch.Tensor): mask from previous pruning iteration
+
+            Returns:
+                new_mask (torch.Tensor): new mask that combines the effects
+                    of the old mask and the new mask from the current 
+                    pruning method (of same dimensions as mask and t).
+            """
+            new_mask = mask  # start off from existing mask
+            new_mask = new_mask.to(dtype=t.dtype)
+
+            # compute a slice of t onto which the new pruning method will operate
+            if method.PRUNING_TYPE == "unstructured":
+                # prune entries of t where the mask is 1
+                slc = mask == 1
+
+            # for struct pruning, exclude channels that have already been
+            # entirely pruned
+            elif method.PRUNING_TYPE == "structured":
+                if not hasattr(method, "axis"):
+                    raise AttributeError(
+                        "Pruning methods of PRUNING_TYPE "
+                        '"structured" need to have the attribute `axis` defined.'
+                    )
+
+                # find the channels to keep by removing the ones that have been
+                # zeroed out already (i.e. where sum(entries) == 0)
+                n_dims = len(t.shape)  # "is this a 2D tensor? 3D? ..."
+                axis = method.axis
+                if axis == -1:
+                    axis = n_dims - 1  # -1 means last dimension
+                # find channels along dim = axis that aren't already tots 0ed out
```

Comment:
You can't remove something that has already been removed. When you try to prune again, you can only consider the remaining channels.
This is especially important for magnitude-based structured pruning: if you remove a channel, in practice, what you do here is not to physically remove it but simply to zero it out (implementation choice); so then, if you try to apply magnitude-based structured pruning again and you _don't_ ignore this pruned channel, then it will always by construction appear to be the lowest magnitude channel, and nothing else will ever get pruned, thus breaking the functionality.

Reply:
but you are changing the random distribution by doing this.

`RandomStructuredPruningMethod` says `Prune entire channels in a tensor at random`. Well, guess what, if you do `RandomStructuredPruningMethod` twice, then you are not pruning the tensor at random, you are using a very different distribution from random.

Either we fix what `RandomStructured` means in the doc, or we fix this code.

Reply:
wait, if I use `RandomStructuredPruningMethod(amount=0.6)` and then again `RandomStructuredPruningMethod(amount=0.2)`, isn't the pattern that emerges equivalent to pruning (60% + (20% of 40%)) = 68% of channels at random? 

In any case, okay, it's clear that at least the wording is misleading. This prunes entire _unpruned_ channels in a tensor at random. Would that make it clear enough?

Reply:
it's whether the `random` here means `Prune entire channels in a tensor at random` or it means `Prune entire non-zero channels in a tensor at random`.

Right now it says `Prune entire channels in a tensor at random`, and by that definition, it has to be `60% of X, and then independently 20% of X`. If you go to the extreme, what if you draw a channel randomly to just be all zeros, then it's considered pruned, though it actually isn't.

Reply:
Good feedback on the wording. Having worked with iterative pruning a lot, I just assumed it was clear that the later pruning rounds would only apply to unpruned channels, so it's good to hear that that's not the case and you would have assumed a completely different behavior. I now specify "(currently unpruned)" everywhere in the docs. 

> If you go to the extreme, what if you draw a channel randomly to just be all zeros, then it's considered pruned, though it actually isn't.

This won't happen because we check the mask to see if a tensor is pruned, not the weights. So, even if the weights are randomly all zeroes, the mask would still be all ones until you actually prune those weights (see `_combine_masks` in `PruningContainer`).